### PR TITLE
add possibility to have user defined callbacks within the training lo…

### PR DESCRIPTION
**Pull request by albertogaspar**
…op of the prob model

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [X] Other (please describe):

## What is the current behavior?

When using `ProbModel`, a user is not able to add any custom action at different stages of the training loop.

## What is the new behavior?

Allow the user to add custom actions at different stages of the training loop, namely:
  - after each gradient update
  - at the beginning of each epoch
  - at the end of each epoch

This behavior is implemented by means of user defined callback as this is the solution adopted by other DL libraries such as [Keras](https://keras.io/api/callbacks/) or [PyTorch Lightning](https://lightning.ai/docs/pytorch/stable/extensions/callbacks.html).

In order to train a `ProbModel` using callbacks the user have to:
- define its own callbacks by subclassing `fortuna.training.callbacks.Callback` and override the method of interest.
- when calling the `train` method for a `ProbModel` instance, add a list of callbacks containing the ones previously defined when initializing `FitConfig`.

Example (assuming to have already initialized a `ProbModel`):
```python
import logging

from jax.flatten_util import ravel_pytree
import optax

from fortuna.prob_model.fit_config import FitConfig, FitMonitor, FitOptimizer
from fortuna.metric.classification import accuracy
from fortuna.training.callbacks import Callback

logger = logging.getLogger(__name__)


class CountParamsCallback(Callback):
    def training_epoch_start(self, state: TrainState) -> TrainState:
        params, unravel = ravel_pytree(state.params)
        logger.info(f"num params: {len(params)}")
        return state


status = prob_model.train(
    train_data_loader=train_data_loader,
    val_data_loader=val_data_loader,
    calib_data_loader=val_data_loader,
    fit_config=FitConfig(
        monitor=FitMonitor(metrics=(accuracy,)),
        optimizer=FitOptimizer(method=optax.adam(1e-4), n_epochs=100),
        processor=FitProcessor(devices=0, disable_jit=False),
        callbacks=[CountParamsCallback()]
    )
)
```

## Other information

In the PR there are also two minor changes not related to the callbacks system:
1. In `normalizing_flow_trainer.py` the typing was not consistent so it was fixed.
2. In `trainer.py` the `Trainer`'s method `training_loss_step` was not available. This method must be implemented in all childs as it contains the logic for the forward pass. Given that `Trainer` is an abstract base class it seems natural to add  `training_loss_step` as an abstract method.

### CODE REVIEW

File: fortuna/training/trainer.py [hunk 1]
```diff
@@ -297,7 +326,7 @@ def _training_loop(
                 training_kwargs,
             )
             # compute training losses and metrics for the current batch
-            training_losses_and_metrics_current_batch = self.training_step_end(
+            state, training_losses_and_metrics_current_batch = self.training_step_end(
```
Comment (gianlucadetommaso) [inline fortuna/training/trainer.py:329]: Shouldn't this have `callbacks` as an argument too?

**Review by gianlucadetommaso — COMMENTED** [review]
Great addition! 

One thing that is missing is the documentation for how to using the callback functionality. In particular, we need to describe this in the API reference part of the documentation (see [here](https://aws-fortuna.readthedocs.io/en/latest/references/prob_model/fit_config.html)).

Could you add some description there, and perhaps include the example you have in this PR? Thanks!
